buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java

```diff
@@ -0,0 +1,187 @@
+        output.append("\n\n");
+
+        // This is a map rather than a set so that the sort order is the 3rd
+        // party component names, unaffected by the full path to the various files
+        final Map<String, File> seen = this.getFilesToAppend(licensesDirs);
+
+        for (Map.Entry<String, File> entry : seen.entrySet()) {
+            final String name = entry.getKey();
+            final File file = entry.getValue();
+            final File licenseFile = new File(file.getParentFile(), name + "-LICENSE.txt");
+
+            appendFileToOutput(file, name, "NOTICE", output);
+            appendFileToOutput(licenseFile, name, "LICENSE", output);
+        }
+
+        write(outputFile,output.toString(),StandardCharsets.UTF_8);
+    }
+
+    private static void appendFileToOutput(File file, final String name, final String type,
+                                           StringBuilder output) throws IOException {
+        String text = readFileToString(file,StandardCharsets.UTF_8);
+        if (text.trim().isEmpty() == false) {
+            output.append("================================================================================\n");
+            output.append(name + " " + type + "\n");
+            output.append("================================================================================\n");
+            output.append(text);
+            output.append("\n\n");
+        }
+    }
+
+    private Map<String,File> getFilesToAppend(List<File> licensesDirectories) throws IOException{
+        final Map<String,File> licensesSeen = new TreeMap<>();
+
+        for (File directory: licensesDirectories) {
+            try(DirectoryStream<Path> stream = Files.newDirectoryStream(directory.toPath())){
+                for (Path path : stream){
+                    if (Files.isRegularFile(path) && path.toString().endsWith("-NOTICE.txt")){
+                        File licenseFile = path.toFile();
+
+                        final String name =
+                            licenseFile.getName().substring(0, licenseFile.getName().length() - "-NOTICE.txt".length());
+
+                        if (licensesSeen.containsKey(name)) {
+                            File prevLicenseFile = licensesSeen.get(name);
+
+                            if (readFileToString(prevLicenseFile,StandardCharsets.UTF_8)
+                                .equals(readFileToString(licenseFile,StandardCharsets.UTF_8)) == false) {
+                                    throw new RuntimeException("Two different notices exist for dependency '" +
+                                        name + "': " + prevLicenseFile + " and " + licenseFile);
+                            }
+                        } else {
+                            licensesSeen.put(name, licenseFile);
+                        }
+                    }
+                }
+            }
+        }
+        return licensesSeen;
+    }
```

Comment:
I ran into issues with using the NIO Files methods using the UTF-8 charsets failing when they were reading the notice files and would run into a "java.nio.charset.MalformedInputException: Input length = 1" when running assemble command using Gradle. So I used the decoder here to cause it to replace the bad characters with the expected bad characters for UTF-8 so it can be clearly seen what characters were replaced.